pkg/cloudprovider/providers/aws/volumecache.go

```diff
@@ -0,0 +1,265 @@
+/*
+Copyright 2014 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aws
+
+import (
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/cloudprovider"
+	"sync"
+)
+
+// volumeCache is a cache of AWS volumes.  See instanceCache for an explanation of the caching approach.
+type volumeCache struct {
+	ec2 EC2
+
+	mutex  sync.RWMutex
+	latest *volumeCacheSnapshot
+}
+
+// volumeCacheSnapshot holds a view at a moment in time of all the volumes
+type volumeCacheSnapshot struct {
+	timestamp  int64
+	byVolumeId map[awsVolumeID]*cachedVolume
+}
+
+// cachedInstance is a reusable volume wrapper, holding some immutable state and caching access to mutable state
+type cachedVolume struct {
+	ID awsVolumeID
+
+	// Values that cannot change are exported publicly
+	AvailabilityZone string
+
+	cache *volumeCache
+
+	// mutex protects the mutable elements below
+	mutex sync.Mutex
+
+	// state is the state in EC2 as observed last
+	state *ec2.Volume
+
+	// stateTimestamp is the timestamp at which we last updated state
+	stateTimestamp int64
+}
+
+// DescribeInstance gets the state for the volume, subject to the cache policy.
+// Currently a cache-miss will cause a refresh of all volumes, because this is the same "price" in terms of quota.
+func (i *cachedVolume) DescribeVolume(cachePolicy *CachePolicy) (*ec2.Volume, error) {
+	snapshot := i.cache.latestSnapshot()
+
+	if snapshot != nil {
+		instance := snapshot.findCachedVolume(i.ID)
```

Comment:
what is the purpose of this findCachedVolume? a cachedVolume must exist, otherwise this function is not going to be called?